python/xgrammar/cuda/apply_token_mask_inplace.py

```diff
@@ -0,0 +1,220 @@
+            nvrtc.nvrtcCreateProgram(
+                str.encode(_apply_token_bitmask_inplace_kernel), b"sourceCode.cu", 0, [], []
+            )
+        )
+        CUDA_HOME = os.getenv("CUDA_HOME")
+        if CUDA_HOME == None:
+            CUDA_HOME = os.getenv("CUDA_PATH")
+        if CUDA_HOME == None:
+            raise RuntimeError("Environment variable CUDA_HOME or CUDA_PATH is not set")
+        include_dirs = os.path.join(CUDA_HOME, "include")
+
+        # Initialize CUDA
+        checkCudaErrors(cudart.cudaFree(0))
+
+        major = checkCudaErrors(
+            cudart.cudaDeviceGetAttribute(
+                cudart.cudaDeviceAttr.cudaDevAttrComputeCapabilityMajor, device_id
+            )
+        )
+        minor = checkCudaErrors(
+            cudart.cudaDeviceGetAttribute(
+                cudart.cudaDeviceAttr.cudaDevAttrComputeCapabilityMinor, device_id
+            )
+        )
+        prefix = "compute"
+        arch_arg = bytes(f"--gpu-architecture={prefix}_{major}{minor}", "ascii")
+
+        try:
+            opts = [
+                b"--fmad=true",
+                arch_arg,
+                "--include-path={}".format(include_dirs).encode("UTF-8"),
+                b"--std=c++11",
+                b"-default-device",
+            ]
+            checkCudaErrors(nvrtc.nvrtcCompileProgram(prog, len(opts), opts))
+        except RuntimeError as err:
+            logSize = checkCudaErrors(nvrtc.nvrtcGetProgramLogSize(prog))
+            log = b" " * logSize
+            checkCudaErrors(nvrtc.nvrtcGetProgramLog(prog, log))
+            # NOTE: The prints below are intended to provide the kernel compilation error.
+            print(log.decode())
+            print(err)
+            raise RuntimeError("CUDA kernel compilation failure")
+
+        dataSize = checkCudaErrors(nvrtc.nvrtcGetPTXSize(prog))
+        data = b" " * dataSize
+        checkCudaErrors(nvrtc.nvrtcGetPTX(prog, data))
+
+        # Store into `_module` and `_func`.
+        module = checkCudaErrors(cuda.cuModuleLoadData(np.char.array(data)))
+        func = checkCudaErrors(cuda.cuModuleGetFunction(module, b"ApplyTokenBitmaskInplaceKernel"))
+        cls._module = module
+        cls._func = func
+        # Return the compiled kernel.
+        return func
+
+
+def apply_token_bitmask_inplace(logits: torch.Tensor, bitmask: torch.Tensor):
+    if cuda is None or cudart is None or nvrtc is None:
+        raise RuntimeError("cuda-python is not installed. Please install cuda-python first.")
+
+    # Check input tensor shapes.
+    if logits.ndim == 2:
+        batch_size, vocab_size = logits.shape
+    elif logits.ndim == 1:
+        batch_size = 1
+        (vocab_size,) = logits.shape
+    else:
+        raise ValueError(f"Invalid logits tensor shape {logits.shape}")
+    bitmask_size = (vocab_size + BITS_PER_BLOCK - 1) // BITS_PER_BLOCK
+
+    # Ensure that the tensors are contiguous in memory.
+    logits = logits.contiguous()
```

Comment:
need to check is_contiguous(), and raise an error if not. This is needed because logits is applied inplace